Update Yices to version 2.8.0

daniel-raffler opened the pull request:
Hello,
this PR updates Yices to version 2.7.0 and closes #402. Other than that the new version mainly includes improvements to the MCSat solver, which we're not using at the moment. [Here](https://github.com/SRI-CSL/yices2/releases/tag/yices-2.7.0) is a full changelog.

Yices2 has some additional features that we may want to include:
- Nonlinear arithmetic is supported with the MCSat solver
- MCSat also support interpolation ([link](https://yices.csl.sri.com/doc/context-operations.html#c.yices_check_context_with_interpolation)), but not `unsat core` ([yet?](https://github.com/SRI-CSL/yices2/pull/578))
- Dpll (which we currently use) has `unsat core`, but no interpolation or non-linear arithmetic
- Only Dpll can be compiled with the `--thread-safety` option that allows multiple solver instances to run in parallel
- Arrays have been added separately, see [here](https://github.com/sosy-lab/java-smt/pull/523) for the other pull request
- Yices can use Cadical/Kissat/CryptoMiniSat as backend for SAT solving. This could be significantly faster, but is currently limited to `QF_BV` logic

In this PR I stuck with Dpll as solver and just enabled the `--thread-safety` option. We could still add MCSat support and maybe even the external SAT solvers. Yices then picks the solver based on the logic, which would give us the best feature set. The big downside is that MCSat is not compatible with `--thread-safety`, which means that multiple parallel solver instances are not possible

**Update**

The thread-safety issues in MCSAT have been fixed, and unsat core is now also supported in MCSAT. We therefore switched to MCSAT as the default solver
- A new option `solver.yices2.solverType=dpllt` has been added to use dpllt instead
- Interpolation is now supported, thanks to the MCSAT solver
- Nonlinear arithmetics and division by zero are also working now

We also switched the entire backend to the official Java bindings for Yices. In addition, a new build script has been added, and Windows binaries are now available

This PR is currently waiting on https://github.com/SRI-CSL/yices2/pull/614, but should be ready to merge otherwise

**Linked issues:**
- https://github.com/SRI-CSL/yices2/issues/608
- https://github.com/SRI-CSL/yices2/issues/613
- https://github.com/SRI-CSL/yices2/issues/615
- https://github.com/SRI-CSL/yices2/issues/616
- https://github.com/SRI-CSL/yices2_java_bindings/pull/8

daniel-raffler committed:
Yices: Add yices_has_mcsat to the bindings

daniel-raffler committed:
Yices: Use absolute paths in the build script

daniel-raffler committed:
Yices: Use the right "version" property when building yices

daniel-raffler committed:
Yices: API updates

daniel-raffler committed:
Yices: Update dependencies for yices

daniel-raffler committed:
Yices: Use yices_get_term_name to get the name of a function

Otherwise Yices may put SMTLIB quotes around the name

daniel-raffler committed:
Yices: Enable tests that require parallel solver instances

When compiled with --thread-safety Yices should be reentrant

daniel-raffler committed:
Yices: Enable VariableNamesTest.testBoolVariableDump()

The issue has been fixed in the new version and "(" characters in variable names now get quoted

daniel-raffler committed:
Yices: Disable tests that are too slow

daniel-raffler committed:
YIces: Fix handling of solver errors

kfriedberger reviewed (commented):
Overall looks good. I will make a deeper test the next days.

daniel-raffler committed:
Merge remote-tracking branch 'origin/master' into yices-2.7.0

# Conflicts:
#	lib/ivy.xml
#	src/org/sosy_lab/java_smt/solvers/yices2/Yices2FormulaCreator.java
#	src/org/sosy_lab/java_smt/solvers/yices2/Yices2Model.java
#	src/org/sosy_lab/java_smt/test/SolverConcurrencyTest.java
#	src/org/sosy_lab/java_smt/test/VariableNamesTest.java

daniel-raffler committed:
Add missing import

daniel-raffler committed:
Yices2: Use install paths in the compilation script

daniel-raffler committed:
Yices2: Update build instructions

daniel-raffler committed:
Merge branch 'master' into yices-2.7.0

# Conflicts:
#	src/org/sosy_lab/java_smt/test/ProverEnvironmentTest.java
#	src/org/sosy_lab/java_smt/test/SolverConcurrencyTest.java
#	src/org/sosy_lab/java_smt/test/SolverFormulaWithAssumptionsTest.java

daniel-raffler committed:
Yices: Remove 32bit build option

daniel-raffler committed:
Yices: Update build script for MCSAT

daniel-raffler committed:
Yices: Update build instructions

daniel-raffler committed:
Yices: Update tests

daniel-raffler committed:
Yices: Use MCSAT in JavaSMT

daniel-raffler committed:
Yices: Add interpolation to JNI bindings

daniel-raffler committed:
Fix editing mistake

daniel-raffler committed:
Yices: Recompile with both thread safety and MCSAT support

Despite what the website says, both options appear to be compatible. However, there are some issues, possibly linked to unsat core, where Yices fails to terminate

daniel-raffler committed:
CI

daniel-raffler committed:
Yices: Rename Yices2TheoremProver to Yices2AbstractProver

daniel-raffler committed:
Yices: Disable array tests

daniel-raffler committed:
Yices: Split off new Yices2Prover class

daniel-raffler committed:
Yices: Add interpolation

daniel-raffler committed:
Yices: Print error message if interpolation fails

daniel-raffler committed:
Yices: Remove build files when `ant clean` is run

daniel-raffler committed:
Yices: Change version name to 2.7.0

daniel-raffler committed:
Yices: Simplify interpolation code

daniel-raffler committed:
Yices: Add a native test to check the solver version

daniel-raffler committed:
Yices: Skip pushing assertions when planning to calculate an unsat core

daniel-raffler committed:
Yices: Re-enable unsat core tests

daniel-raffler committed:
Yices: Use dpllt for unsat core

daniel-raffler committed:
Yices: Switch back to MCSAT for all provers

daniel-raffler committed:
Yices: Short circuit formula translation if the other context is also Yices

Yices has no equivalent to SolverContext and stores all its terms globally. There is no need to move terms from one SolverContext to another, and the original term can simply be used in the new SolverContext

daniel-raffler committed:
Yices: Fix assumption tracking

daniel-raffler committed:
Yices: Switch to using the official Java bindings

See https://github.com/SRI-CSL/yices2_java_bindings

daniel-raffler committed:
Yices: Update tests for Yices 2.8

daniel-raffler committed:
Yices: Add support for sums and products in the visitor

daniel-raffler committed:
Yices: Fix bv sums

daniel-raffler committed:
Yices: Print error message when function is called with wrong number of arguments

daniel-raffler committed:
Yices: Suppress `unchecked` warnings

daniel-raffler committed:
Yices: Use larger print window to make sure formulas are not cut off

daniel-raffler committed:
Use a simpler example in `simpleInterpolation` and `emptyInterpolationGroup` test

The old example is too complex for several solvers now (cvc5, yices2, opensmt)

daniel-raffler committed:
Merge branch 'refs/heads/master' into yices-2.7.0

# Conflicts:
#	lib/ivy.xml

daniel-raffler committed:
Yices: Fix handling of UFs with no arguments

daniel-raffler committed:
Yices: Add an option for picking the solver (DPLLT or MCSat)

daniel-raffler committed:
Yices: Add a hack to handle bv terms in the visitors

Yices rewrites bv terms into boolean operations. We should restore the original term for the visitor to handle this correctly

daniel-raffler committed:
Yices: Actually apply solver options

daniel-raffler committed:
Yices: Update build instructions

daniel-raffler committed:
Yices: Add a build script

daniel-raffler committed:
Yices: Add yices patch

daniel-raffler committed:
Yices: Remove dead code

daniel-raffler committed:
Yices: Turn missing native tests back on

daniel-raffler committed:
Yices: Update tests

Segfaults have been fixed by https://github.com/SRI-CSL/yices2/pull/614

daniel-raffler committed:
Yices: Update more model tests

daniel-raffler committed:
Yices: Check shutdown notifier before interpolating, and print a better error message if interpolation fails

daniel-raffler committed:
Yices: Turn a model test back on

The test still takes several seconds, but no longer seems to hang

daniel-raffler committed:
Yices: Update build docs

daniel-raffler committed:
Format

daniel-raffler committed:
Merge branch 'refs/heads/master' into yices-2.7.0

# Conflicts:
#	src/org/sosy_lab/java_smt/solvers/yices2/Yices2AbstractProver.java
#	src/org/sosy_lab/java_smt/solvers/yices2/Yices2FormulaCreator.java
#	src/org/sosy_lab/java_smt/solvers/yices2/Yices2NativeApi.java

daniel-raffler committed:
Clean up

daniel-raffler committed:
Yices: Add comments about failing interpolation tests

daniel-raffler committed:
Yices: Add support for arrays with multiple indices

daniel-raffler committed:
Yices: Set `hasPersistentModel` to true

daniel-raffler committed:
Yices: Make new classes package private

daniel-raffler committed:
Yices: Use assumption solving wrapper

This fixes some corner cases where interpolation isn't available after isUnsatWithAssumptions

daniel-raffler committed:
Yices: Fix issue in getModel when GENERATE_UNSAT_CORE has been set

daniel-raffler committed:
Yices: Inline a method

daniel-raffler committed:
Yices: Add missing newline

daniel-raffler committed:
Yices: Turn the final interpolation test back on

The test will hang when using the Intelij runner, but work fine when using `ant tests`

daniel-raffler committed:
Yices: Use property to skip loading the Yices library

daniel-raffler committed:
Yices: Add ant targets to clean the build

daniel-raffler committed:
Yices: Use generic methods to handle sum and product terms

daniel-raffler committed:
Yices: Add Windows binaries

daniel-raffler committed:
Yices: Only use 8 threads for the mingw build to avoid OOM crashes

daniel-raffler committed:
Yices: Update feature table

daniel-raffler committed:
Yices: Disable autoloader in native tests

daniel-raffler committed:
Yices: Enable Windows tests

daniel-raffler committed:
Yices: Enable division by zero tests

daniel-raffler committed:
CI

daniel-raffler committed:
Yices: Fix an error message

daniel-raffler committed:
Yices: Add support for modulo in the visitor

daniel-raffler committed:
Yices: Fix products in the visitor

daniel-raffler committed:
Yices: Fix products/sums with more than 2 terms

daniel-raffler committed:
Yices: Rewrite "bit" and "bool-to-bv" terms in the visitor

daniel-raffler committed:
Yices: Don't rewrite and/or terms with more than 2 arguments

daniel-raffler committed:
Yices: Update maven script

kfriedberger committed:
Merge remote-tracking branch 'origin/master' into yices-2.7.0

ThomasHaas commented:
> Yices has github actions ([link](https://github.com/SRI-CSL/yices2/blob/master/.github/actions/build/action.yml)) to build the solver on MacOS. We could probably start from there if we also want to support that platform

I would love Yices2 support on MacOS (ARM). It usually outperforms Z3 (and others) by quite a lot on our QF_BV encoding in Dartagnan (sometimes by a whole order of magnitude). It seems to do some serious black magic there :)

kfriedberger committed:
Yices2: fix unit test for Boolector and CVC4 (Linux only).

kfriedberger committed:
Yices2: update build-scripts and documentation

daniel-raffler commented:
Hello Thomas,

I've been able to build the MacOS (Arm) binaries with a github runner. We still have to figure out to sign and ship the binaries, but I can at least include the build script. It's not too complicated  to run and the binaries can then just be copied over

> > Yices has github actions ([link](https://github.com/SRI-CSL/yices2/blob/master/.github/actions/build/action.yml)) to build the solver on MacOS. We could probably start from there if we also want to support that platform
> 
> I would love Yices2 support on MacOS (ARM). It usually outperforms Z3 (and others) by quite a lot on our QF_BV encoding in Dartagnan (sometimes by a whole order of magnitude). It seems to do some serious black magic there :)

Is this with the default solver, or would we need one of [these](https://web.archive.org/web/20251212142346/https://yices.csl.sri.com/doc/install-sources.html#third-party-sat-solvers) delegates? So far we only have the internal solver, but we might still add one of the delegates

kfriedberger committed:
Yices2: update JavaSMT bindings for Yices2 to v6.0.0-141-g04134287c

kfriedberger committed:
Yices2: apply Refaster findings.

kfriedberger committed:
Yices2: ignore certain architectures or older systems in tests.

kfriedberger committed:
Yices2: fix AppVeyor CI config.

disteph commented:
> > Is there a plan to integrate https://github.com/daniel-raffler/yices2_java_bindings.git into its original? Just asking, not urgent.
> 
> Yes, I've open a PR last week: [SRI-CSL/yices2_java_bindings#8](https://github.com/SRI-CSL/yices2_java_bindings/pull/8)

Your PR is merged! I've also aligned the Windows JNI with it (with a test harness including windows).

disteph commented:
News: SAT-solver delegates are now available in Yices contexts, including support for push/pop behavior, but still limited to QF_BV logic:
https://github.com/SRI-CSL/yices2/pull/607

These other issues have been fixed and merged into master:
https://github.com/SRI-CSL/yices2/issues/608
https://github.com/SRI-CSL/yices2/issues/613
https://github.com/SRI-CSL/yices2/issues/615

The last one remaining in the issues listed in this thread is 
https://github.com/SRI-CSL/yices2/issues/616
I can't reproduce, but I have a tentative fix/hardening at https://github.com/SRI-CSL/yices2/pull/628

daniel-raffler commented:
> News: SAT-solver delegates are now available in Yices contexts, including support for push/pop behavior, but still limited to QF_BV logic: [SRI-CSL/yices2#607](https://github.com/SRI-CSL/yices2/pull/607)
> 
> These other issues have been fixed and merged into master: [SRI-CSL/yices2#608](https://github.com/SRI-CSL/yices2/issues/608) [SRI-CSL/yices2#613](https://github.com/SRI-CSL/yices2/issues/613) [SRI-CSL/yices2#615](https://github.com/SRI-CSL/yices2/issues/615)
> 
> The last one remaining in the issues listed in this thread is [SRI-CSL/yices2#616](https://github.com/SRI-CSL/yices2/issues/616) I can't reproduce, but I have a tentative fix/hardening at [SRI-CSL/yices2#628](https://github.com/SRI-CSL/yices2/pull/628)

Thanks for the update! I ran some benchmarks and the segfaults during interpolation are completely gone now. We still get exceptions for these tasks, but at least that's something we can react to, and maybe try a different strategy. Note there are still a few unrelated crashes that don't appear to be fixed by https://github.com/SRI-CSL/yices2/pull/628 either. However, the issue is incredibly rare (< 4 in 18000 tasks), so it might be fine to just ignore it for now

Good work!

disteph commented:
> > News: SAT-solver delegates are now available in Yices contexts, including support for push/pop behavior, but still limited to QF_BV logic: [SRI-CSL/yices2#607](https://github.com/SRI-CSL/yices2/pull/607)
> > These other issues have been fixed and merged into master: [SRI-CSL/yices2#608](https://github.com/SRI-CSL/yices2/issues/608) [SRI-CSL/yices2#613](https://github.com/SRI-CSL/yices2/issues/613) [SRI-CSL/yices2#615](https://github.com/SRI-CSL/yices2/issues/615)
> > The last one remaining in the issues listed in this thread is [SRI-CSL/yices2#616](https://github.com/SRI-CSL/yices2/issues/616) I can't reproduce, but I have a tentative fix/hardening at [SRI-CSL/yices2#628](https://github.com/SRI-CSL/yices2/pull/628)
> 
> Thanks for the update! I ran some benchmarks and the segfaults during interpolation are completely gone now. We still get exceptions for these tasks, but at least that's something we can react to, and maybe try a different strategy. Note there are still a few unrelated crashes that don't appear to be fixed by [SRI-CSL/yices2#628](https://github.com/SRI-CSL/yices2/pull/628) either. However, the issue is incredibly rare (< 4 in 18000 tasks), so it might be fine to just ignore it for now
> 
> Good work!

Now merged into master. I've pushed another commit fixing another data race; maybe it addresses the remaining 4 tasks. Feel free to rerun your tests on master.
